File: 品味Java的21種設計模式/dp_java/builder/example3/TxtBuilder.java
package cn.javass.dp.builder.example3;

import java.util.Collection;
import java.util.Map;

/**
 * 实现导出数据到文本文件的的构建器对象
 */
public class TxtBuilder implements Builder {
	/**
	 * 用来记录构建的文件的内容，相当于产品
	 */
	private StringBuffer buffer = new StringBuffer();
	public void buildBody(
			Map<String, Collection<ExportDataModel>> mapData) {
		for(String tblName : mapData.keySet()){
			//先拼接表名称
			buffer.append(tblName+"\n");
			//然后循环拼接具体数据
			for(ExportDataModel edm : mapData.get(tblName)){
				buffer.append(edm.getProductId()+","+edm.getPrice()+","+edm.getAmount()+"\n");
			}
		}
	}
	public void buildFooter(ExportFooterModel efm) {
		buffer.append(efm.getExportUser());
	}
	public void buildHeader(ExportHeaderModel ehm) {
		buffer.append(ehm.getDepId()+","+ehm.getExportDate()+"\n");
	}	
	public StringBuffer getResult(){
		return buffer;
	}	
}

File: 品味Java的21種設計模式/dp_cpp/builder/example3/TxtBuilder.cpp
#include "TxtBuilder.h"

namespace cn
{
	namespace javass
	{
		namespace dp
		{
			namespace builder
			{
				namespace example3
				{

					void TxtBuilder::buildBody(Map<std::string, Collection<ExportDataModel*>*> *mapData)
					{
						for (Map<std::string, Collection<ExportDataModel*>*>::const_iterator tblName = mapData->begin(); tblName != mapData->end(); ++tblName)
						{
							//先拼接表名称
							buffer->append(tblName->first+"\n");
							//然后循环拼接具体数据
							for (unknown::const_iterator edm = mapData->get(tblName).begin(); edm != mapData->get(tblName).end(); ++edm)
							{
								buffer->append((*edm)->getProductId()+","+(*edm)->getPrice()+","+(*edm)->getAmount()+"\n");
							}
						}
					}

					void TxtBuilder::buildFooter(ExportFooterModel *efm)
					{
						buffer->append(efm->getExportUser());
					}

					void TxtBuilder::buildHeader(ExportHeaderModel *ehm)
					{
						buffer->append(ehm->getDepId()+","+ehm->getExportDate()+"\n");
					}

					StringBuffer *TxtBuilder::getResult()
					{
						return buffer;
					}
				}
			}
		}
	}
}
